File: 113. 路径总和 II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    public int requireSum;
    public List<List<Integer>> PathsList =new LinkedList<>();
    public List<Integer> currentPathList=new LinkedList<>();
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        requireSum=sum;
        preOrder(root,0);
        return PathsList;
    }
    
    public void preOrder(TreeNode root,int currentSum){
        if(root!=null){
            currentSum+=root.val;
            currentPathList.add(root.val);
            if(root.left==null&&root.right==null){
                if(currentSum==requireSum){
                    PathsList.add(currentPathList);
                }
                currentPathList.remove(currentPathList.size()-1);
            }
            preOrder(root.left,currentSum);
            preOrder(root.right,currentSum);
        }
    }
}

File: js/113. 路径总和 II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    if(!root)
        return []
    let ans=[]
    let res=[]
    const traverse=function(root,preSum,ifLeaf,depth){

        ifLeaf=!(root.left||root.right)
        ans[depth]=root.val
        
        if(ifLeaf&&preSum+root.val===sum){
            res.push(ans.slice(0,depth+1))
        }

        root.left&&traverse(root.left,preSum+root.val,ifLeaf,depth+1)
        root.right&&traverse(root.right,preSum+root.val,ifLeaf,depth+1)
    }
    traverse(root,0,!(root.left||root.right),0)
    return res
};
